x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java

```
@@ -0,0 +1,126 @@
+    private final ThreadPool threadPool;
+    private final TransportService transportService;
+    private final AsyncSearchStoreService store;
+
+    @Inject
+    public TransportGetAsyncSearchAction(TransportService transportService,
+                                         ActionFilters actionFilters,
+                                         ClusterService clusterService,
+                                         NamedWriteableRegistry registry,
+                                         Client client,
+                                         ThreadPool threadPool) {
+        super(GetAsyncSearchAction.NAME, transportService, actionFilters, GetAsyncSearchAction.Request::new);
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.threadPool = threadPool;
+        this.store = new AsyncSearchStoreService(taskManager, threadPool, client, registry);
+    }
+
+    @Override
+    protected void doExecute(Task task, GetAsyncSearchAction.Request request, ActionListener<AsyncSearchResponse> listener) {
+        try {
+            AsyncSearchId searchId = AsyncSearchId.decode(request.getId());
+            if (clusterService.localNode().getId().equals(searchId.getTaskId().getNodeId())) {
+                getSearchResponseFromTask(request, searchId, listener);
+            } else {
+                TransportRequestOptions.Builder builder = TransportRequestOptions.builder();
+                DiscoveryNode node = clusterService.state().nodes().get(searchId.getTaskId().getNodeId());
+                if (node == null) {
+                    getSearchResponseFromIndex(request, searchId, listener);
+                } else {
+                    transportService.sendRequest(node, GetAsyncSearchAction.NAME, request, builder.build(),
+                        new ActionListenerResponseHandler<>(listener, AsyncSearchResponse::new, ThreadPool.Names.SAME));
+                }
+            }
+        } catch (IOException e) {
```

should we catch all exceptions?

---

It shouldn't be necessary but I pushed https://github.com/elastic/elasticsearch/pull/49931/commits/9056b873039b5f087a7f98dee1fd31d3850dde11 to be on the safe side ;).